src/intersections/boxes.jl

```diff
@@ -34,3 +34,59 @@ function intersecttype(f::Function, b1::Box{Dim,T}, b2::Box{Dim,T}) where {Dim,T
     return FaceTouchingBoxes(Box(u, v)) |> f
   end
 end
+
+"""
+    intersecttype(function, ray, box)
+
+Calculate the intersection type of a `ray`` and a `box`` and apply function `f` to it.
+
+## References
+
+* Williams A, Barrus S, Morley R K, et al., 2005. [An efficient and robust ray-box
+  intersection algorithm]
+  (https://dl.acm.org/doi/abs/10.1145/1198555.1198748)
+"""
+function intersecttype(f::Function, r::Ray{3,T}, b::Box{3,T}) where {T}
+  sign = (x -> x < 0 ? 2 : 1).(direction(r))
+  invdir = one(T) ./ direction(r)
+  bounds = (coordinates(b.min), coordinates(b.max))
+  ori = coordinates(origin(r))
+
+  # Use index access to avoid unnecessary comparitions.
+  # Avoid the following code:
+  # if (invdir[1] >= 0)
+  #   tmin = (coordinates(b.min)[1] - ori[1]) * invdir[1]
+  #   tmax = (coordinates(b.max)[1] - ori[1]) * invdir[1]
+  # else
+  #   tmax = (coordinates(b.min)[1] - ori[1]) * invdir[1]
+  #   tmin = (coordinates(b.max)[1] - ori[1]) * invdir[1]
+  # end
```

Comment:
If the commented code is more intuitive, please use it instead. We can always come back to the optimization with sign, which may not be relevant after all.